apollo-router/src/plugins/telemetry/filtering_layer.rs

```diff
@@ -0,0 +1,141 @@
+        // The logic is splitted between `enabled` and `event_enabled` because we can't evaluate the value of an attribute in this method
+        // This method is evaluated only once for a specific callsite, which means if I return false here I won't re-enter in this method everytime for this event
+        // Which is better for performance. So for supergraph level we can use this method, for subgraph level we have to know the value of subgraph attribute
+        if metadata.level() == &Level::DEBUG && metadata.is_event() {
+            match &self.attributes_to_exclude {
+                Some(attributes_to_exclude) => {
+                    let mut subgraph = false;
+                    let mut to_exclude = false;
+                    for field_name in metadata.fields().iter().map(|f| f.name()) {
+                        if field_name == SUBGRAPH_ATTRIBUTE_NAME {
+                            subgraph = true;
+                            // Cannot do anything for subgraph at this level. It will be in `event_enabled` method.
+                            break;
+                        }
+                        if attributes_to_exclude.supergraph.contains(field_name) {
+                            to_exclude = true;
+                        }
+                    }
+                    if !subgraph && to_exclude {
+                        return false;
+                    }
+                }
+                None => {
+                    if metadata
+                        .fields()
+                        .iter()
+                        .map(|f| f.name())
+                        .any(|n| SPECIFIC_ATTRIBUTES.contains(&n))
+                    {
+                        return false;
+                    }
+                }
+            }
+        }
+
+        ctx.enabled(metadata)
+    }
+
+    fn event_enabled(
+        &self,
+        event: &tracing::Event<'_>,
+        _ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        if event.metadata().level() == &Level::DEBUG {
+            let mut fields_visitor = FieldsVisitor::default();
+            event.record(&mut fields_visitor);
+            let field_names: HashSet<&str> = fields_visitor.fields.keys().copied().collect();
+            // Check only for subgraphs here because other cases have been catched before in `enabled`
+            if let Some(attributes_to_exclude) = &self.attributes_to_exclude {
+                if let Some(subgraph) = fields_visitor.fields.get(SUBGRAPH_ATTRIBUTE_NAME) {
+                    if let Some(attrs_to_exclude) = attributes_to_exclude
+                        .subgraphs
+                        .get(subgraph)
+                        .map(|a| a.iter().map(|a| a.as_str()).collect::<HashSet<&str>>())
+                    {
+                        if !attrs_to_exclude.is_disjoint(&field_names) {
+                            return false;
+                        }
+                    } else if !attributes_to_exclude
```

Comment:
Implementation consideration: Why maintain `all_subgraphs` as a backstop? Why not de-normalise the `all_subgraphs` values into each individual `subgraph`? May take slightly more memory, but would simplify and possibly speed up processing.

Reply:
Hm I'm not sure it would speed up processing, maybe simplify yes. But it means we need to access to global configuration or global parsed supergraph to know what's the list of subgraphs. I don't think we have a clean implementation of this currently but that might be something we should provide in the `PluginInit` struct ?

Reply:
I deleted the `DEBUG` level condition in the process